### PR TITLE
Add statistics about answer frequency

### LINKED ISSUE #852: Session question view: add statistics about answer frequency

### PR DESCRIPTION
**Pull request by levinuss**
This adds toggleable stats about answer choice frequency by previous users to the SessionAnswerView. The stats are re-calculated every hour for every question that had an answer submitted in the meantime.

Closes #852

### CODE REVIEW

File: resources/js/SessionCardAnswerView.svelte [hunk 1]
```diff
@@ -22,6 +23,9 @@
     <div class="row border-light border-start border-3 m-1 p-2">
         <p class="p-1">{@html DOMPurify.sanitize(answer.text)}</p>
     </div>
+    {#if settingsShowAnswerStats}
+        <p class="mt-2 badge bg-secondary" title="Percentage of users who knew this answer">{answer.answer_percentage}%</p>
```
Comment (schu) [inline resources/js/SessionCardAnswerView.svelte:27]: If `settingsShowAnswerStats` is enabled but `answer_precentage` hasn't been calcuclated yet for this answers, it would display `null%`. I.e. it's missing a check `answer.answer_percentage != null` as done in SessionAnswerView.

But I don't think there's much value in adding this info to card questions, so I would prefer if we don't calculate/add it for card questions at all. (Also, I think "users who knew this" isn't necessarily true, since it's primarily a flag if the user wants to see the card again or not and not necessarily a reflection of what they actually know..?)

File: app/Models/AnswerChoice.php [hunk 2]
```diff
@@ -11,7 +11,7 @@ class AnswerChoice extends Model
 
     protected $fillable = ['question_id', 'answer_id', 'is_correct', 'session_id', 'help_used'];
 
-    public function questions()
+    public function question()
```
Comment (schu) [inline app/Models/AnswerChoice.php:14]: 👍

File: resources/js/UserSettingsView.svelte [hunk 3]
```diff
@@ -43,6 +43,10 @@
         <input class="form-check-input" type="checkbox" role="switch" bind:checked={settings.session_multiple_answer_tries}>
         <label class="form-check-label" for="flexSwitchCheckDefault">Allow selecting multiple answers: the correct answer remains hidden until it is selected</label>
     </div>
+    <div class="form-check form-switch">
+        <input class="form-check-input" type="checkbox" role="switch" bind:checked={settings.session_show_answer_stats}>
+        <label class="form-check-label" for="flexSwitchCheckDefault">Show statistics about answer choice frequency by previous users</label>
```
Comment (schu) [inline resources/js/UserSettingsView.svelte:48]: Let's just say `Show answer distribution stats` or something along the lines.

File: app/Console/Kernel.php [hunk 4]
```diff
@@ -104,6 +105,60 @@ protected function schedule(Schedule $schedule): void
             Cache::put('stats/decks/last_used', $lastUsedDecks);
         })->everyTwoMinutes();
 
+        $schedule->call(function () {
+            // Get the 10 answer_choices that were answered next since the last answer_choice that was processed
+            // or start the queue by using the last 100 answer_choices if there is no id in the cache.
+            // Also, get the related questions, their answers and the answer_choices_count of both.
+            $idLastCalculatedAnswerChoice = Cache::get('internal/last_answer_percentage_calculation');
+
+            $answer_choices = AnswerChoice::select('id', 'question_id', 'answer_id')
```
Comment (schu) [inline app/Console/Kernel.php:114]: Hmm, that's a pretty complex query and hard to parse and understand.

Can we simplify it by having two queries, depending on if we have a previous ID or not? Why not take 10 in both cases?

Nit: variable name should be camelCase.

File: app/Console/Kernel.php [hunk 5]
```diff
@@ -104,6 +105,60 @@ protected function schedule(Schedule $schedule): void
             Cache::put('stats/decks/last_used', $lastUsedDecks);
         })->everyTwoMinutes();
 
+        $schedule->call(function () {
+            // Get the 10 answer_choices that were answered next since the last answer_choice that was processed
+            // or start the queue by using the last 100 answer_choices if there is no id in the cache.
+            // Also, get the related questions, their answers and the answer_choices_count of both.
+            $idLastCalculatedAnswerChoice = Cache::get('internal/last_answer_percentage_calculation');
```
Comment (schu) [inline app/Console/Kernel.php:112]: Nit: I would prefer a shorter variable name, for example `lastAnswerChoiceID`. Since everything happens in the scope of a few line of code, the context should be clear nonetheless.

File: app/Console/Kernel.php [hunk 6]
```diff
@@ -104,6 +105,60 @@ protected function schedule(Schedule $schedule): void
...
+                    $answerChoicesCount = $answer->answer_choices_count;
+                    $answerPercentage = $answerChoicesCount / $questionAnswerChoicesCount * 100;
+                    $roundedPercentage = round($answerPercentage);
+                    $answer->answer_percentage = $roundedPercentage;
+                    $totalPercentage += $answerPercentage;
+                    $totalRoundedPercentage += $roundedPercentage;
+                    Log::debug('total percentage at ' . $answer->id . ': ' . $totalPercentage);
+
+                    // If this is the last answer and the total rounded percentage is not 100, adjust the last answer
+                    if (round($totalPercentage) == 100) {
```
Comment (schu) [inline app/Console/Kernel.php:146]: Should this be `!= 100` instead?

File: app/Console/Kernel.php [hunk 7]
```diff
@@ -104,6 +105,60 @@ protected function schedule(Schedule $schedule): void
...
+                        Log::debug('total percentage is 100');
+                        $difference = 100 - $totalRoundedPercentage;
+                        $answer->answer_percentage += $difference;
+                        $totalRoundedPercentage += $difference;
+                    }
+                    $answer->save();
+                }
+                if ($idLastCalculatedAnswerChoice < $answer_choice->id) {
```
Comment (schu) [inline app/Console/Kernel.php:154]: Can we take `anwer_choices` in ascending order and avoid the check?

**Review by schu — CHANGES_REQUESTED** [review]

File: app/Console/Kernel.php (hunk 4 above)
Comment (levinuss) [reply app/Console/Kernel.php:114]: I agree for the two queries.
The 100 in case there is no id in the cache is because I wanted to include the last 100 answer choices in the new queue (but still only process the oldest 10 of them this time). This way, the last 100 answer choices will be processed in the next runs of the scheduler. But the value can definitely be discussed.

File: resources/js/SessionCardAnswerView.svelte (hunk 1 above)
Comment (levinuss) [reply resources/js/SessionCardAnswerView.svelte]: I removed the stats for card questions.

**Comment by schu** [comment]
Thanks!
